File: Medium/2044. Count Number of Maximum Bitwise-OR Subsets.py
from ast import List

class Backtracking:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        max_or = 0

        for num in nums:
            max_or |= num

        self.cnt = 0

        self.backtrack(nums, 0, 0, max_or)
        return self.cnt
    
    def backtrack(self, nums: List[int], idx: int, cur_or: int, max_or: int) -> None:
            if cur_or == max_or:
                self.cnt += 1 << (len(nums) - idx)
                return
            for i in range(idx, len(nums)):
                self.backtrack(nums, i + 1, cur_or | nums[i], max_or)
                
class Counter:
    def countMaxOrSubsets(self, nums: List[int]) -> int:
        max_or = 0
        prev = Counter()
        prev[0] = 1

        for num in nums:
            max_or |= num
            curr = Counter()
            for prev_or, cnt in prev.items():
                curr[prev_or | num] += cnt
            prev.update(curr)
        
        return prev[max_or]

File: Medium/2044. Count Number of Maximum Bitwise-OR Subsets.java
import java.util.HashMap;
import java.util.Map;

/**
 * Backtracking solution to count the number of subsets with maximum bitwise OR.
 */
class Backtracking {
    private int count = 0;

    /**
     * Calculates the number of subsets whose bitwise OR is maximized.
     *
     * @param nums input array of integers
     * @return count of subsets with maximum OR
     */
    public int countMaxOrSubsets(int[] nums) {
        int maxOr = 0;
        // Compute the maximum possible OR value for the array
        for (int num : nums) {
            maxOr |= num;
        }
        // Start backtracking from index 0, with current OR as 0
        backtrack(nums, 0, 0, maxOr);
        return count;
    }

    /**
     * Recursively explores all subset combinations.
     * If the current OR equals the maximum, all remaining elements can be
     * included/excluded freely,
     * so we add 2^(n-idx) to the count.
     */
    private void backtrack(int[] nums, int idx, int currOr, int maxOr) {
        if (currOr == maxOr) {
            count += 1 << (nums.length - idx);
            return;
        }
        for (int i = idx; i < nums.length; i++) {
            backtrack(nums, i + 1, currOr | nums[i], maxOr);
        }
    }
}

/**
 * Enumeration (dynamic programming) solution to count the number of subsets
 * with maximum bitwise OR.
 */
class Enumeration {
    /**
     * Calculates the number of subsets whose bitwise OR is maximized.
     *
     * @param nums input array of integers
     * @return count of subsets with maximum OR
     */
    public int countMaxOrSubsets(int[] nums) {
        int maxOr = 0;
        // prev maps OR value to the number of ways to achieve it
        Map<Integer, Integer> prev = new HashMap<>();
        prev.put(0, 1); // Start with the empty subset

        for (int num : nums) {
            maxOr |= num;
            Map<Integer, Integer> curr = new HashMap<>();
            // For each previous OR value, add the current number to the subset
            for (Map.Entry<Integer, Integer> entry : prev.entrySet()) {
                int newOr = entry.getKey() | num;
                curr.put(newOr, curr.getOrDefault(newOr, 0) + entry.getValue());
            }
            // Merge current results back into prev for the next iteration
            for (Map.Entry<Integer, Integer> entry : curr.entrySet()) {
                prev.put(entry.getKey(), prev.getOrDefault(entry.getKey(), 0) + entry.getValue());
            }
        }
        // Return the number of ways to achieve the maximum OR
        return prev.getOrDefault(maxOr, 0);
    }
}